### ISSUE TITLE
Explore knime: what can/cannot do

### ISSUE THREAD

**Issue by jiqicn** [EyeofBeholder-NLeSC/knime-demo#1]
# General
- Available on Windows, Linux, and Mac
- Pricing
  - KNIME Analytics Platform is open source and free to use locally
  - KNIME Server needs to be paid
- The [KNIME Forum](https://forum.knime.com/) 
- Allow [Python integration](https://www.knime.com/blog/how-to-setup-the-python-extension)
  - Extensions are developed in Java
  - Support running Jupyter inside KNIME (see this [link](https://hub.knime.com/mpattadkal/spaces/Public/latest/Jupyter%20Webinar/01_Run%20Jupyter%20in%20KNIME/Run%20Jupyter%20in%20KNIME~6rzm1Wn3Zsr4NO7-))
- Node monitor allows closer looking at the runtime situation of nodes.
- KNIME allows more flexible workflow control than RapidMiner. For instance, it allows to add nodes for configuring other nodes by passing a **flow variable** to those nodes, and this can be applied to all the children nodes on the same branch (need further setup). KNIME also supports **if switch** that allows the user to choose which branch the workflow should go.
- Nodes can be grouped and aggregated as a metanode/component to clean up mess workflow. The difference from these two are minor. Basically, a component has more functions and ports than a meta node, see the table below:
  <img width="520" alt="image" src="https://user-images.githubusercontent.com/92043159/170000263-3ac442be-7654-4905-8603-7473d333ca44.png">

**Comment by jiqicn** [comment]
# Visualization and exploration of the input dataset

## Data visualization as a dashboard

Built-in visualization nodes that are already very useful. To generate a dashboard that contains multiple linked visualization, it's possible to aggregate them together as a component and linked the desired variables. This will generate a dashboard that contains all the visualizations included. In this case, visualizations are automatically linked by the data attrs shared between the nodes (it requires the relevant options in the "Interactivity" panel to be checked).

<img width="627" alt="image" src="https://user-images.githubusercontent.com/92043159/170005926-de4c69c8-cc85-47b3-b6b2-e0518220ef84.png">

It's also possible to customize the layout of the dashboard:

<img width="619" alt="image" src="https://user-images.githubusercontent.com/92043159/170006510-0a900530-ed89-42ce-aa99-85d21070c8c5.png">

There are also some very nice nodes for visual data exploration, for instance, the official component of automated visualization:

<img width="860" alt="image" src="https://user-images.githubusercontent.com/92043159/169779064-4befa616-5b96-4c6c-9380-fac8c86aeada.png">

## Script integration for visualization
The "Python View" node helps integrating Python code to generate static visualization of data using Python visualization libraries such as matplotlib and seaborn. Also, a surprisingly nice feature is the Generic JavaScript View, which allows us to create highly customized visualization in JavaScript. It supports some widely-used and well-maintained libraries, such as D3 and Plotly. Something even better is that the Generic JavaScript View can be combined with some other built-in JS nodes to make a dashboard. I'm still looking for a way to allow the general JS view node to subscribe/publish events (this [link](https://forum.knime.com/t/how-to-publish-subscribe-events-using-custom-javascript-views/18705/5) might be useful).

<img width="230" alt="image" src="https://user-images.githubusercontent.com/92043159/170006357-41616050-3270-481f-9a49-91af81841a33.png">

<img width="637" alt="image" src="https://user-images.githubusercontent.com/92043159/168789241-411271b4-7ae7-4221-89c2-67a90ab7f3dc.png">

## Visualization control
Some of the visualization nodes are intractable, which means the users could choose what attribute(s) they want to visualize. But for most of the view nodes implemented in JS, it doesn't support any control on it. In this case, it is possible to add a widget node that is connected to the key flow variable of the views. By checking the re-execution option, it will refresh the linked views after changing the value of the widget.

<img width="319" alt="image" src="https://user-images.githubusercontent.com/92043159/170062048-f08bad98-5710-4dcf-a605-106c263ac852.png">

**Comment by jiqicn** [comment]
# Train new models and choose a good pipeline

## Pre-processing
The political statement dataset is stored as a csv file, and that can be easily loaded by dragging-and-dropping. There are nodes for converting statement strings to documents, so that the tagging node could work on the data. 

Different taggers are supported, for instance, the POS tagger and the Stanford NE tagger. It also support to tag documents by using a pre-defined dictionary (with dictionary tagger). Steaming, case converting, and filtering are all possible with the corresponding nodes.

<img width="542" alt="image" src="https://user-images.githubusercontent.com/92043159/169997279-912b4ece-164d-4e47-9f2f-9467752d3515.png">

## Transformation
To transform the statement documents to a form that can be used for training machine learning models, a typical way of doing this is to vectorize the documents somehow. KNIME supports different ways of word/doc vectorization, e.g. TFIDF and word2vec.

<img width="491" alt="image" src="https://user-images.githubusercontent.com/92043159/169998157-f03995d5-d692-455d-a545-d723a49097ce.png">

## Training machine learning models
Similar to RapidMiner that various machine learning models are directly available as nodes. Also, you can download nodes or components that implement other nodes that are not included by default, e.g.,  XGBoost and deep neural networks.

Typically, a model contains two parts, which are the model leaner and predictor. The leaner node takes the training set as input for training the model, while the predictor node will test the performance of the model on the testing set. An example that trains a C4.5 by cross validation:

<img width="646" alt="image" src="https://user-images.githubusercontent.com/92043159/169999485-d52a807c-6c41-4004-a6ec-c9b92f81dbd4.png">

The performance of the model can be evaluated through a separate node, e.g., scorer, numeric scorer, etc., where some metrics (confusion matrix, accuracy, error, etc.) will be summarized and presented together.

<img width="282" alt="image" src="https://user-images.githubusercontent.com/92043159/170000987-e3a2c083-9e90-4cb3-a588-2860dcacb3bd.png">

**Comment by jiqicn** [comment]
# Explain training results
## Built-in methods in KNIME
KNIME natively supports two methods/algorithms for explaining prediction results of machine learning models (binary/multinomial): shapely additive explanation and LIME. 

Both methods support creating an explanation on a single data item, by providing scores to the variables to indicate how much that variable supports/opposes the classification. Visualization could be something like this:

![image](https://user-images.githubusercontent.com/92043159/171874867-3d20a7bf-25b7-4ca9-926d-8626f98346b8.png)

The left panel shows the information of the item (preprocessed document and tokens with TF.IDF value), and the right bar chart shows the results of SHAP and LIME separately. It's kind of weird that the two methods explain the results in totally opposite ways, so that's something that needs to be integrated further.
